### PR TITLE
✨ add `metrics` endpoints to fastapi apps for promethius scraping

### LINKED ISSUE #5216: Add `metric` endpoints to all simcore services to be scraped by prometheus
### Is there an existing issue for this?

- [X] I have searched the existing issues

### User Story

Once we release we need to be able to diagnose issues in simcore services, in order to improve the quality of the platform. One way to do this is by using Prometheus. For that each service needs a `metrics` endpoint which exposes diagnostics.

### Definition of Done

All simcore services have a `metrics` endpoint which exposes data to prometheus and Grafana dashboard(s) have been setup in order to get an easy overview of the services.

### PR DESCRIPTION
**Pull request by bisgaard-itis**
## What do these changes do?
- move https://github.com/trallnag/prometheus-fastapi-instrumentator dependency to `servicelib[fastapi]`
- expose `/metrics` endpoint for prometheus scraping and instrument the following fastapi services:
  - agent
  - autoscaling
  - catalog
  - clusters-keeper
  - datcore-adapter
  - director-v2
  - dynamic-scheduler
  - invitations
  - payments
  - resource-usage-tracker

## Related issue/s

- closes #5216 

## Dev Checklist

- [x] No ENV changes or I properly updated ENV ([read the instruction](https://git.speag.com/oSparc/osparc-ops-deployment-configuration/-/blob/configs/README.md?ref_type=heads#how-to-update-env-variables))

### CODE REVIEW

**Review by sanderegg — APPROVED** [review]
already love it! thanks a bunch!

**Review by GitHK — APPROVED** [review]
👍 Very nice.

**Review by matusdrobuliak66 — APPROVED** [review]
:sunglasses:

File: packages/service-library/src/servicelib/fastapi/prometheus_instrumentation.py [hunk 1]
```diff
@@ -0,0 +1,23 @@
+# pylint: disable=protected-access
+
+
+from fastapi import FastAPI
+from prometheus_fastapi_instrumentator import Instrumentator
+
+
+def instrument_app(app: FastAPI):
```
Comment (pcrespov) [inline packages/service-library/src/servicelib/fastapi/prometheus_instrumentation.py:8]: I would rename this as `setup_prometheus_instrumentation(app: FastAPI)`

File: packages/service-library/src/servicelib/fastapi/prometheus_instrumentation.py [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+    instrumentator = (
+        Instrumentator(
+            should_instrument_requests_inprogress=True, inprogress_labels=False
+        )
+        .instrument(app)
+        .expose(app, include_in_schema=False)
+    )
+
+    def _unregister():
+        for collector in list(instrumentator.registry._collector_to_names.keys()):
```
Comment (pcrespov) [inline packages/service-library/src/servicelib/fastapi/prometheus_instrumentation.py:19]: You do not need `list` here. actually you can directly iterate as
```python
for collector in instrumentator.registry._collector_to_names:
   ...
```

File: services/agent/requirements/_base.in [hunk 3]
```diff
@@ -8,7 +8,7 @@
 # intra-repo required dependencies
 --requirement ../../../packages/models-library/requirements/_base.in
 --requirement ../../../packages/settings-library/requirements/_base.in
---requirement ../../../packages/service-library/requirements/_base.in
+--requirement ../../../packages/service-library/requirements/_fastapi.in
```
Comment (pcrespov) [inline services/agent/requirements/_base.in:11]: thx!

I would have done this in a separate PR but I guess the agent is not critical

File: services/agent/src/simcore_service_agent/core/settings.py [hunk 4]
```diff
@@ -46,6 +46,7 @@ class ApplicationSettings(BaseCustomSettings, MixinLoggingSettings):
     AGENT_VOLUMES_CLEANUP_INTERVAL_S: NonNegativeInt = Field(
         60 * _MINUTE, description="interval at which to repeat volumes cleanup"
     )
+    AGENT_ADD_METRICS_ENDPOINT: bool = True
```
Comment (pcrespov) [inline services/agent/src/simcore_service_agent/core/settings.py:49]: TIP: I would rather call it `*_PROMETHEUS_INSTRUMENTATION_ENABLED`

File: packages/service-library/src/servicelib/fastapi/prometheus_instrumentation.py [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+
+
+from fastapi import FastAPI
+from prometheus_fastapi_instrumentator import Instrumentator
+
+
+def instrument_app(app: FastAPI):
+
+    instrumentator = (
+        Instrumentator(
```
Comment (pcrespov) [inline packages/service-library/src/servicelib/fastapi/prometheus_instrumentation.py:11]: Q: What kind of metrics is this instrumentator collecting? 
- In aiohttp we had a lower level interface and we had to specify sensors etc ...
- Moreover, we had issues with cardinality that would totally overload the system.
- Have you tested the impact of the reading on `/metric` on the system  (NOTE: in the past these handlers were sync and therefore blocking over-time )

Comment (bisgaard-itis) [reply packages/service-library/src/servicelib/fastapi/prometheus_instrumentation.py:11]: I think I saw that we in the `aiohttp` use https://pypi.org/project/prometheus-client/. The package I use here builds on top of that to have a simple interface for fastapi. 
Concerning which metrics are collected, you can see that [here](https://github.com/trallnag/prometheus-fastapi-instrumentator#prometheus-fastapi-instrumentator-). In addition to the default ones I collect the number of requests currently in progress, which I believe is very useful. 
I have not tested the performance impact, but the developer makes [this](https://github.com/trallnag/prometheus-fastapi-instrumentator#disclaimer) disclaimer about the package. However, my point of view is that we need to collect some metrics and I would prefer to use a tool rather than define the lowlevel metrics myself. So until I see this one causes issues I think we should go for it.

File: packages/service-library/src/servicelib/fastapi/prometheus_instrumentation.py (hunk 1 above)
Comment (bisgaard-itis) [reply packages/service-library/src/servicelib/fastapi/prometheus_instrumentation.py:8]: fixed

File: packages/service-library/src/servicelib/fastapi/prometheus_instrumentation.py (hunk 2 above)
Comment (bisgaard-itis) [reply packages/service-library/src/servicelib/fastapi/prometheus_instrumentation.py:19]: done

**Review by pcrespov — APPROVED** [review]
